### PR TITLE
Run each integration test in their own process

### PR DESCRIPTION
**Pull request by mikker**
Skips `spec/integration` by default. Instead it runs then separately in each their own spec.

It's very hard to not have the Rails specs bleed into each other if we want to test different behaviours with different setups. This way each integration test has its very own, close-to pristine universe.

### CODE REVIEW

File: bin/run-tests [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+base="bundle exec rspec -f progress -f JUnit -o spec/ruby-agent-junit.xml"
```
Comment (mikker) [inline bin/run-tests]: @kuisathaverat We probably want to separate the junit outputs too but I'm unsure what's possible in Jenkins.

**Comment by mikker** [comment]
Only test failure is one of the custom jruby images failing to build `http-parser` (dep via `http`). Let's hear if the automation folks can figure out what's up.

File: spec/scripts/spec.sh [hunk 2]
```diff
@@ -43,4 +43,4 @@ RUBY_VERSION=${VERSION} docker-compose run \
   /bin/bash -c "\
     gem install rake && \
     bundle install && \
-    timeout -s9 5m bundle exec rspec -f progress -f JUnit -o spec/ruby-agent-junit.xml ${TEST}"
+    timeout -s9 5m bin/run-tests ${TEST}"
```
Comment (kuisathaverat) [inline spec/scripts/spec.sh]: I do not see a reason to split the script, it calls another script that is in the folder bin, it is weird

Comment (mikker) [reply spec/scripts/spec.sh]: Did you see the script?

Comment (mikker) [reply spec/scripts/spec.sh]: It needs to run _one_ command when passed a `$1` and multiple if not.

File: bin/run-tests [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+set -x
+
+base="bundle exec rspec -f progress -f JUnit -o spec/ruby-agent-junit.xml"
+specific_spec=$1
+
+if [[ $specific_spec = '' ]]; then
+  echo 'Running all specs, including integration'
+
+  $base
+  find spec/integration -name '*_spec.rb' -exec $base {} \;
```
Comment (kuisathaverat) [inline bin/run-tests]: it launches several tests in a sequential way, each one will generate a junit file, if all use the same name we will only have the latest one, so we have to make the name uniq for each test, we would need a function

Comment (kuisathaverat) [inline bin/run-tests]:
```suggestion
  find spec/integration -name '*_spec.rb' -exec runBaseCase {} \;
```

File: bin/run-tests [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+set -x
+
+base="bundle exec rspec -f progress -f JUnit -o spec/ruby-agent-junit.xml"
+specific_spec=$1
+
+if [[ $specific_spec = '' ]]; then
+  echo 'Running all specs, including integration'
+
+  $base
```
Comment (kuisathaverat) [inline bin/run-tests]:
```suggestion
  runBaseCase
```

File: bin/run-tests [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+set -x
+
+base="bundle exec rspec -f progress -f JUnit -o spec/ruby-agent-junit.xml"
```
Comment (kuisathaverat) [inline bin/run-tests]:
```suggestion
runBaseCase(){
  local case=${1:-""}
  bundle exec rspec -f progress -f JUnit -o spec/ruby-agent-junit${case}.xml ${case}
}
```

File: bin/run-tests [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+if [[ $specific_spec = '' ]]; then
+  echo 'Running all specs, including integration'
+
+  $base
+  find spec/integration -name '*_spec.rb' -exec $base {} \;
+else
+  echo "Running only $specific_spec"
+
+  $base $1
```
Comment (kuisathaverat) [inline bin/run-tests]:
```suggestion
  runBaseCase $1
```

File: spec/scripts/spec.sh (hunk 2 above)
Comment (kuisathaverat) [reply spec/scripts/spec.sh]: we can use a function

File: spec/elastic_apm/spies/mongo_spec.rb [hunk 7]
```diff
@@ -53,32 +56,32 @@ def start_mongodb
...
+      with_agent do
+        client =
+          Mongo::Client.new(
+            [url],
+            database: 'elastic-apm-test',
+            logger: Logger.new(nil),
+            server_selection_timeout: 5
+          )
+
+        # ParallelCollectionScan can only be run on an existing collection.
```
Comment (estolfo) [inline spec/elastic_apm/spies/mongo_spec.rb]: This comment can be removed now.

**Comment by mikker** [comment]
@kuisathaverat I don't see how your function makes anything different? The proposal format makes the changes a bit hard to follow.

File: spec/elastic_apm/spies/mongo_spec.rb [hunk 8]
```diff
@@ -89,28 +92,25 @@ def start_mongodb
       expect(db.type).to eq 'mongodb'
       # ParallelCollectionScan doesn't send 'lsid' in the command so we can
       # validate the entire command document.
-      expect(db.statement).to eq '{"parallelCollectionScan"=>"testing", ' \
-        '"numCursors"=>1}'
+      expect(db.statement).to match('"delete"=>"testing"')
```
Comment (estolfo) [inline spec/elastic_apm/spies/mongo_spec.rb]: You can remove the comment about ParallelCollectionScan but I still don't understand how you can validate the entire delete command? When I was writing this test, the `lsid` value was in the command so I couldn't use delete and chose ParallelCollectionScan instead. Is there some setup somewhere I'm not aware of?

**Comment by mikker** [comment]
I got it on the second read. Thanks! Did you see the build failure on Jenkins?

File: spec/elastic_apm/spies/mongo_spec.rb (hunk 8 above)
Comment (estolfo) [reply spec/elastic_apm/spies/mongo_spec.rb]: For example, I get this `"{\"delete\"=>\"testing\", \"ordered\"=>true, \"deletes\"=>[{\"q\"=>{}, \"limit\"=>0}], \"lsid\"=>{\"id\"=><BSON::Binary:0x70210864279520 type=uuid data=0x4c4b83c656b94211...>}}"`

**Comment by kuisathaverat** [comment]
yep, I am on it it is failing on jRuby 9.2-12 only should be somethignon this image or in that version I have to check it, probably I ask for help I'm not a ruby guy :)

File: spec/elastic_apm/spies/mongo_spec.rb (hunk 8 above)
Comment (mikker) [reply spec/elastic_apm/spies/mongo_spec.rb]: `match(String)` (as opposed to `eq()`) uses `include?` under the hood, I think, so it's just a matter of the string being in there somewhere.

The reason I changed it at all was that my Mongo (that I finally figured out had moved to their own homebrew tap) was telling me that `parallelCollectionScan` had been removed.

Comment (estolfo) [reply spec/elastic_apm/spies/mongo_spec.rb]: Oh great, I didn't notice you used `match` instead. This works too!

**Comment by kuisathaverat** [comment]
jenkins, run the tests

**Comment by mikker** [comment]
omg @kuisathaverat this is so gloriously green

green-light
